spec/lib/mastodon/cli/maintenance_spec.rb

```diff
@@ -4,9 +4,64 @@
 require 'mastodon/cli/maintenance'
 
 describe Mastodon::CLI::Maintenance do
+  let(:cli) { described_class.new }
+
   describe '.exit_on_failure?' do
     it 'returns true' do
       expect(described_class.exit_on_failure?).to be true
     end
   end
+
+  describe '#fix_duplicates' do
+    context 'when the database version is too old' do
+      before do
+        allow(ActiveRecord::Migrator).to receive(:current_version).and_return(2000_01_01_000000) # Earlier than minimum
+      end
+
+      it 'Exits with error message' do
+        expect { cli.invoke :fix_duplicates }.to output(
+          a_string_including('is too old')
+        ).to_stdout.and raise_error(SystemExit)
+      end
+    end
+
+    context 'when the database version is too new and the user does not continue' do
+      before do
+        allow(ActiveRecord::Migrator).to receive(:current_version).and_return(2100_01_01_000000) # Later than maximum
+        allow(cli.shell).to receive(:yes?).with('Continue anyway? (Yes/No)').and_return(false).once
+      end
+
+      it 'Exits with error message' do
+        expect { cli.invoke :fix_duplicates }.to output(
+          a_string_including('more recent')
+        ).to_stdout.and raise_error(SystemExit)
+      end
+    end
+
+    context 'when Sidekiq is running' do
+      before do
+        allow(ActiveRecord::Migrator).to receive(:current_version).and_return(2022_01_01_000000) # Higher than minimum, lower than maximum
+        allow(Sidekiq::ProcessSet).to receive(:new).and_return [:process]
+      end
+
+      it 'Exits with error message' do
+        expect { cli.invoke :fix_duplicates }.to output(
+          a_string_including('Sidekiq is running')
+        ).to_stdout.and raise_error(SystemExit)
+      end
+    end
+
+    context 'when the database is too new and the user does continue', use_transactional_tests: false do
+      before do
+        allow(cli.shell).to receive(:yes?).with('Continue anyway? (Yes/No)').and_return(true).once # Database too new query
+        allow(cli.shell).to receive(:yes?).with('Continue? (Yes/No)').and_return(true).once # General task warning
+      end
+
+      it 'removes duplicate records' do
+        expect { cli.invoke :fix_duplicates }.to output(
+          a_string_including('Finished')
+        ).to_stdout
+      end
+    end
```

Comment:
> I'm not too sure about testing that, because it messes with the database schema in a pretty specific way and the “Database too new” thing is there for good reasons, so I'm not sure testing that really makes a lot of sense.

I think what I was handling there is that in the current state of the repo, the constant in this file is older than the latest migration, so in order to run this script you have to answer yes to this question.

Should that constant be updated to reflect that it can run correctly on current schema/migration/version? ... or should I update this spec to instead stub out the DB version?

> It's also difficult testing the behavior of the script because the script makes the most sense in cases where _database consistency assumptions are broken_.

Yes - compared to the other CLI scripts this has been more challenging to wrap my head around how its interacting with the app and how to mock or not those interactions.

Reply:
The constant in this file is older than the latest migration because that script is not super well maintained. Its primary reason to exist is to handle issues caused by a libc migration back in 2018. Similar issues can arise again but they are less likely, and it is most likely that people will run into the issue when running earlier migrations well within the range supported by the script.